crates/punktf-lib/src/visit/diff/mod.rs

```diff
@@ -0,0 +1,210 @@
+//! A [`Visitor`](`crate::visit::Visitor`) implementation which creates events for
+//! files which differ from the content it would have once deployed.
+
+use crate::{
+	profile::LayeredProfile,
+	profile::{source::PunktfSource, transform::Transform},
+	visit::*,
+};
+use std::path::Path;
+
+/// Applies any relevant [`Transform`](`crate::profile::transform::Transform`)
+/// for the given file.
+fn transform_content(profile: &LayeredProfile, file: &File<'_>, content: String) -> String {
+	let mut content = content;
+
+	// Copy so we exec_dotfile is not referenced by this in case an error occurs.
+	let exec_transformers: Vec<_> = file.dotfile().transformers.to_vec();
+
+	// Apply transformers.
+	// Order:
+	//   - Transformers which are specified in the profile root
+	//   - Transformers which are specified on a specific dotfile of a profile
+	for transformer in profile.transformers().chain(exec_transformers.iter()) {
+		content = transformer.transform(content).unwrap();
+	}
+
+	content
+}
+
+/// An event which is emitted for every differing item.
+#[derive(Debug)]
+pub enum Event<'a> {
```

Comment:
could we prevent livetimes by doing the same to Path as we did to &str? I am sure there is a owned variant of path.

Reply:
alternatively use &str, but having both owned and unowned seems wrong to me?

Reply:
We could also borrow the strings, but they are not used anywhere else, so why not transfer ownership.

The paths, instead are owned by other structs which will want to keep them. So we could not move them out. The only way would be to `Clone` them but that might not be needed in most cases. And if ownership over the path is needed in a downstream implementer, they are free to call `Clone`.

I would vote to keep it this way.

(fyi: The owned Variant of `Path` is `PathBuf`).